arangosh/Restore/RestoreFeature.cpp

```diff
@@ -1306,65 +967,534 @@ arangodb::Result processInputDirectory(
-    // special case: never restore data in the _users collection first as it could
-    // potentially change user permissions. In that case index creation will fail.
-    res = ::restoreIndexes(httpClient, jobData);
-    if (res.ok()) {
-      res = ::restoreData(httpClient, jobData);
+  try {
+    res = job.run(client);
+  } catch (arangodb::basics::Exception const& ex) {
+    res.reset(ex.code(), ex.what());
+  } catch (std::exception const& ex) {
+    res.reset(TRI_ERROR_INTERNAL, ex.what());
+  } catch (...) {
+    res.reset(TRI_ERROR_INTERNAL, "unknown exception");
+  }
+
+  if (res.fail()) {
+    job.feature.reportError(res);
+  }
+}
+
+
+}  // namespace
+
+namespace arangodb {
+
+RestoreFeature::RestoreJob::RestoreJob(RestoreFeature& feature,
+                                       RestoreProgressTracker& progressTracker,
+                                       RestoreFeature::Options const& options,
+                                       RestoreFeature::Stats& stats, 
+                                       bool useEnvelope,
+                                       std::string const& collectionName,
+                                       std::shared_ptr<SharedState> sharedState)
+    : feature{feature}, 
+      progressTracker{progressTracker}, 
+      options{options}, 
+      stats{stats}, 
+      useEnvelope{useEnvelope},
+      collectionName(collectionName),
+      sharedState(std::move(sharedState)) {}
+
+RestoreFeature::RestoreJob::~RestoreJob() = default;
+
+Result RestoreFeature::RestoreJob::sendRestoreData(arangodb::httpclient::SimpleHttpClient& client,
+                                                   size_t readOffset,
+                                                   char const* buffer,
+                                                   size_t bufferSize) {
+  using arangodb::basics::StringUtils::urlEncode;
+  using arangodb::httpclient::SimpleHttpResult;
+  
+  std::string const url = "/_api/replication/restore-data?collection=" + urlEncode(collectionName) +
+                          "&force=" + (options.force ? "true" : "false") +
+                          "&useEnvelope=" + (useEnvelope ? "true" : "false");
+  
+  std::unordered_map<std::string, std::string> headers;
+  headers.emplace(arangodb::StaticStrings::ContentTypeHeader,
+                  arangodb::StaticStrings::MimeTypeDump);
+
+  std::unique_ptr<SimpleHttpResult> response(
+      client.request(arangodb::rest::RequestType::PUT, url, buffer, bufferSize, headers));
+  arangodb::Result res = ::checkHttpResponse(client, response, "restoring data", "");
+  
+  if (res.fail()) {
+    // error
+    LOG_TOPIC("a595a", WARN, Logger::RESTORE)
+        << "Error while restoring data into collection '" << collectionName
+        << "': " << res.errorMessage();
+    // leave readOffset in place in readOffsets, because the job did not succeed
+    {
+      // store error
+      MUTEX_LOCKER(locker, sharedState->mutex);
+      sharedState->result = res;
```

Comment:
Here we are potentially overwriting previous errors. Should we consider collecting the errors instead so we can report them all? However, this would probably require rather big changes...

Reply:
Yes, we could be overwriting an already stored error with another error.
But we will never overwrite an error with a non-error here.
I don't think it is a problem, because with default startup parameters the restore will stop on an error. The error is also reported to the user. Which error it is does not really matter, and if multi-threading is involved, will likely depend on scheduling and such.
It may be useful to collect all the errors here, but it is a bigger change and I didn't want to do it in context of this PR.